Draft a page on polyfills (#38326)
mdn/content: files/en-us/mdn/writing_guidelines/page_structures/polyfills/index.md

Polyfills are important to web developers but are also a risk: buggy polyfills can break websites or create security vulnerabilities. For this reason, MDN recommends specific sources for polyfills, and is very conservative about adding additional sources.

## Polyfills in the JavaScript reference

### Selected polyfills

The [JavaScript reference documentation](/en-US/docs/Web/JavaScript) can link to polyfills from two sources:

- The [core-js](https://github.com/zloirock/core-js/tree/master) library.
- The [es-shims](https://github.com/es-shims) organization.

Sources are selected by the MDN maintainers based on the following criteria:

- Conformance to the specification for the feature for which they are providing an implementation.
- Adoption by the web development community, as seen in metrics such as the number of [npm](https://www.npmjs.com/) downloads.

> Anyone can propose that MDN should recognize an additional source of polyfills by [filing an issue in the mdn/content repository](https://github.com/mdn/content/issues). However, the MDN maintainers expect the number of recognized polyfills linked from MDN to remain very small, to reduce the risk of recommending polyfills that cause problems for web developers.

Reviewer: I recommend sending issues in mdn/mdn, or mdn/mdn-community, since it rarely only concerns one or a few pages.

Author: You mean, start a discussion in https://github.com/orgs/mdn/discussions? mdn-community doesn't seem to support issues. I would be happy to say that instead.

Reviewer: Oh yeah, a discussion in mdn-community or an issue in mdn/mdn is what I meant.

### Integration of polyfills into pages

> When a page in the JavaScript reference documentation links to a polyfill, it adds the link in the "See also" section at the end of the page.

Reviewer: You can even copy the exact wording we use:

```md
- [Polyfill for `featureName` in `project-name`](link)
```

And mention that they should be at the top of the list

## Polyfills in the Web API reference

### Selected polyfills

> The [Web API reference documentation](/en-US/docs/Web/API) can link to polyfills that are maintained alongside the specification for the feature itself.

Reviewer: This seems a little vague. Maybe a few examples would help.

Author: Hm, well for example the repo that hosts the trusted types specification includes a polyfill: https://github.com/w3c/trusted-types?tab=readme-ov-file#polyfill. Is it enough to add that as an example or do we need to come up with some general form of words for it? "maintained in the same repo as the spec" seems a bit restrictive?

Reviewer: Maybe "maintained by the spec champions"? "Officially endorsed"? I'm not sure how blessed that polyfill is; I just know that TC39 is very cautious not to endorse any polyfill, even one developed by the champions themselves. But in any case yes a link to that page is what I'm looking for at the minimum.

Author: > Maybe "maintained by the spec champions"? "Officially endorsed"?

But (if we're going to be picky) these also seem vague. Who are the spec champions, and do we know that they maintain it, or just that they trust the people who do. And what does official endorsement look like? I mean in the TT case, I'm pretty sure I want MDN to link to it, and I'm pretty sure I trust it, because it's in the same repo as the spec, and I would assume that the maintainers keep that repo under control. I suppose "maintained alongside" is supposed to be a version of "maintained in the same repo as", with a little extra wiggle room in case it ends up in a separate repo.

Reviewer: Alright that sounds reasonable. Linking to that example would still help.

> ---

Reviewer 2: Cross link this page from https://developer.mozilla.org/en-US/docs/Glossary/Polyfill

Reviewer: I don't think we should link to MDN-internal stuff for other pages. I think that would cause some confusion for readers while not being too helpful

Author: Agree with @Josh-Cena here, this is MDN-metadocs, not web docs.

title: Polyfills
slug: MDN/Writing_guidelines/Page_structures/Polyfills
page-type: mdn-writing-guide
sidebar: mdnsidebar
---

> This page outlines MDN's policy for including polyfills in two areas of the reference documentation: the [JavaScript documentation](/en-US/docs/Web/JavaScript), and the [documentation of Web APIs](/en-US/docs/Web/API).

Reviewer 2: ```suggestion
This page outlines MDN's policy for including polyfills in the reference documentation for [JavaScript](/en-US/docs/Web/JavaScript), and [Web APIs](/en-US/docs/Web/API).
```

> On the web, a {{glossary("Polyfill", "polyfill")}} is a piece of code used to provide an implementation of a web platform feature in browsers that do not natively support that feature. Polyfills enable web developers to write a single codebase targeting multiple browsers and browser versions, even when some of those browsers don't support some of the features used.

Reviewer 2: This feels clunky. Perhaps
```suggestion
A {{glossary("Polyfill", "polyfill")}} is an implementation of a web platform feature that websites can use in browsers that do not natively support the feature. Polyfills enable web developers to write a single codebase targeting multiple browsers and browser versions, even when some of those browsers don't support some of the features used.
```

Polyfills are important to web developers but are also a risk: buggy polyfills can break websites or create security vulnerabilities. For this reason, MDN recommends specific sources for polyfills, and is very conservative about adding additional sources.

## Polyfills in the JavaScript reference

### Selected polyfills

The [JavaScript reference documentation](/en-US/docs/Web/JavaScript) can link to polyfills from two sources:

- The [core-js](https://github.com/zloirock/core-js/tree/master) library.
- The [es-shims](https://github.com/es-shims) organization.

Sources are selected by the MDN maintainers based on the following criteria:

- Conformance to the specification for the feature for which they are providing an implementation.
- Adoption by the web development community, as seen in metrics such as the number of [npm](https://www.npmjs.com/) downloads.

> Anyone can propose that MDN should recognize an additional source of polyfills by [starting a discussion in the MDN discussion forum](https://github.com/orgs/mdn/discussions). However, the MDN maintainers expect the number of recognized polyfills linked from MDN to remain very small, to reduce the risk of recommending polyfills that cause problems for web developers.

Reviewer 2: I would give this its own heading.

Author: I split the policy at this level because I think currently our JS policy is much more advanced than our Web/API policy. In fact you can't really call our usage in Web/API a policy at all, I just tried to document our practice as I see it. I don't know if we want to allow "non-spec WG" polyfills for Web/API at all, we haven't even had that conversation, And, again, the idea of this PR is to unblock the situation wrt JavaScript polyfills.

Author: I will add the extra heading though 👍

### Integration of polyfills into pages

When a page in the JavaScript reference documentation links to a polyfill, it adds the link in the "See also" section at the end of the page.

The link is placed at the start of the "See also" list, in the following format:

```md
- [Polyfill for `featureName` in `project-name`](link)
```

> 

Reviewer 2: When should it be removed? My understanding is that we remove the links when there is support in all browsers in the head revision. Needs to be confirmed.

Reviewer: See above ^ My personal vote is to keep them forever, since any responsible software needs to support browsers that are at least a few years old.

Reviewer 2: I have no firm opinion. If ^^^ is agreed, then it is worth spelling out the policy though.

Author: As https://github.com/mdn/content/pull/38326#issuecomment-2683067944 says, I'd prefer to defer these conversations. This PR is just to move things along a bit, so we can even have a documented polyfill  policy, and allow other polyfills than core-js. That alone has taken several years to get done!

sidebar: mdnsidebar
---

> This page outlines MDN's policy for including polyfills in the reference documentation for [JavaScript](/en-US/docs/Web/JavaScript), and [Web APIs](/en-US/docs/Web/API).

Reviewer 2: I think I left a comma in there that I shouldn't have.
